### ISSUE TITLE
Optimise schedule based on past builds

### ISSUE THREAD

**Issue by sgraf812** [ndmitchell/shake#791]
I wondered today whether `shake` already implements the following scheduling idea:

If I had infinitely many Cores, it's pretty obvious that shake can schedule each rule as they become "available". But if I only have e.g. 4 cores, shake must somehow decide which rule to schedule next. Without further knowledge about how much time a rule takes, it will probably prioritise the critical plan of the build plan. But I think it could be a bit smarter, by doing something akin to profile-guided optimisation: Instead of assuming that all jobs take the same time when building the build plan, it could look at profiles of past builds and see "Oh, although module A only has a height of 4 (in the dep graph sorted by height levels), it takes so long to compile that I should probably prioritise it over module B, which has height level 10".

Those profiles are an optional input and could be populated from e.g. CI.

**Comment by ndmitchell** [comment]
It's certainly feasible, and I considered it, but I haven't yet had a case where it seems significant (but I don't have enough information to know for sure). Shake uses random scheduling, which in practice is enough to avoid all worst case behaviours - you can be smarter, but if you are smarter and wrong, it usually goes badly. Random gives you great average case, and for most builds, the average and optimal cases are not that far apart (whereas the worst case can be significantly further behind). It also has reschedule so people who know more can do something smarter, but even the utility of that has been marginal.

Shake already records the information you are talking about in its database, so could do something with it, if we could figure out what exactly the right thing to do is, and measure its performance impact. It's probably that last step that is the hardest - what does a representative benchmark look like? Maybe building the last 1000 commits of GHC through Hadrian incrementally would give you a reasonable approximation.

**Comment by sgraf812** [comment]
I agree, probably nothing worth doing until we notice it in a profile.
